### PR TITLE
Initial implementation of the libp2p introspection server

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For https://github.com/libp2p/go-libp2p/issues/775

This is the first draft of the introspection server.

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
### **TODO**
- [ ] Release go-libp2p-core & configure the version here
- [ ] Unit test the `DefaultIntrospector`
- [ ] CircleCI config & makefile
- [ ] CLI for easy testing of the server

**Review by raulk — COMMENTED** [review]
Would like to see what the fetch logic ends up looking like after integrating my review comments from libp2p/go-libp2p-core#107. Besides that, nothing controversial here so far.

**Comment by aarshkshah1992** [comment]
@raulk  The new fetch logic is up.

**Comment by aarshkshah1992** [comment]
@raulk  Please can you explain how to determine `State.StartTs` & `State.SnapshotDurationMs`

File: introspection/default_introspector.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		// resolve streams on connection
+		if d.tree.Stream != nil {
+			for _, c := range conns {
+				var sids []introspect.StreamID
+				for _, s := range c.Streams.StreamIds {
+					sids = append(sids, introspect.StreamID(s))
+				}
+
+				sl, err := d.tree.Stream(introspect.StreamQueryParams{introspect.QueryOutputFull, sids})
```
Comment (raulk) [inline introspection/default_introspector.go:82]: This is fine for now, but we could further optimise this by collecting all stream IDs from all connections, and fetching them all at once, then merging them into the appropriate connections. Could you file an issue to optimise this?

File: introspection/ws_server.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+var upgrader = websocket.Upgrader{}
+
+// StartServer starts the ws introspection server with the given introspector
+func StartServer(introspector introspect.Introspector) func() error {
+	// introspect handler
+	http.HandleFunc("/introspect", toHttpHandler(introspector))
+
+	// start server
+	serverInstance := http.Server{
+		// TODO Need a better strategy to select an address
```
Comment (raulk) [inline introspection/ws_server.go:24]: Let's make sure this doesn't fall off the radar by opening an issue.

File: introspection/default_introspector.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+var _ introspect.Introspector = (*DefaultIntrospector)(nil)
+
+// DefaultIntrospector is a registry of subsystem data/metrics providers and also allows
+// clients to inspect the system state by calling all the providers registered with it
+type DefaultIntrospector struct {
+	treeMu     sync.RWMutex
+	tree       *introspect.DataProviders
+	listenAddr []string
```
Comment (raulk) [inline introspection/default_introspector.go:21]: Modified this variable name.

File: introspection/ws_server.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	return func() error {
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+
+		return srv.Shutdown(shutdownCtx)
+	}
+}
+
+func wsUpgrader(introspector introspect.Introspector) http.HandlerFunc {
```
Comment (raulk) [inline introspection/ws_server.go:45]: Renamed this method.

File: introspection/ws_server.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			}
+
+			bz, err := proto.Marshal(state)
+			if err != nil {
+				logger.Errorf("failed to marshal introspector state, err: %s", err)
+				return
+			}
+
+			// send the response
+			wsConn.SetWriteDeadline(time.Now().Add(10 * time.Second))
```
Comment (raulk) [inline introspection/ws_server.go:78]: Increased the timeout here.

File: go.mod [hunk 6]
```diff
@@ -0,0 +1,12 @@
+module github.com/libp2p/go-libp2p-introspection
+
+go 1.13
+
+require (
+	github.com/gogo/protobuf v1.3.1
+	github.com/gorilla/websocket v1.4.1
+	github.com/imdario/mergo v0.3.8
+	github.com/ipfs/go-log v1.0.1
+	github.com/libp2p/go-libp2p-core v0.3.1-0.20200206170018-fbc69c15df4c
```
Comment (raulk) [inline go.mod:10]: This import was missing. It's locked to a commit, but it's fine for now.

**Review by raulk — APPROVED** [review]

File: introspection/default_introspector.go (hunk 1 above)
Comment (aarshkshah1992) [reply introspection/default_introspector.go:82]: Done in #5

File: introspection/ws_server.go (hunk 2 above)
Comment (aarshkshah1992) [reply introspection/ws_server.go:24]: Done in https://github.com/libp2p/go-libp2p-introspection/issues/6

**Pull request merged**